### PR TITLE
ISPN-14194 Add config property for default max result

### PR DESCRIPTION
**Pull request by fax4ever**
https://issues.redhat.com/browse/ISPN-14194

~~Just the first step, that is adding the config property for it.~~ ready for the review
I'm writing the documentation

### CODE REVIEW

File: core/src/main/resources/schema/infinispan-config-14.0.xsd [hunk 1]
```diff
@@ -968,6 +968,21 @@
...
+      <xs:element name="query" minOccurs="0">
+        <xs:annotation>
+          <xs:documentation>
+            Defines query options for cache
+          </xs:documentation>
+        </xs:annotation>
+        <xs:attribute name="default-max-result" type="xs:integer" default="${Query.default-max-result}">
+          <xs:annotation>
+            <xs:documentation>
+              Applies the given value as maxResults as a default to any query (indexed, non-indexed and hybrid).
```
Comment (dvagnero) [inline core/src/main/resources/schema/infinispan-config-14.0.xsd:980]: A suggestion:
Limits the number of returned results by a query. Applies to indexed, non-indexed, and hybrid queries. Setting the default maxResult significantly improves performance of queries that don't have an explicit maxResults set.

**Comment by tristantarrant** [comment]
The name of the property should be plural (`default-max-results`)

**Comment by fax4ever** [comment]
Thank you for the review @dvagnero and @tristantarrant, going to apply the changes.
In the meantime I have completed the implementation, but we still need to address the documentation.

**Comment by fax4ever** [comment]
I applied the changes and fixed the last stuff. I'm writing my doc additions.

File: query/src/main/java/org/infinispan/query/impl/QueryDefinition.java [hunk 2]
```diff
@@ -104,6 +111,10 @@ public SearchQueryBuilder getSearchQueryBuilder() {
       return searchQueryBuilder;
    }
 
+   public boolean maxResultsHasChanged() {
```
Comment (tristantarrant) [inline query/src/main/java/org/infinispan/query/impl/QueryDefinition.java:114]: I don't like the name. Maybe `isCustomMaxResults()` ?

File: core/src/main/java/org/infinispan/configuration/cache/QueryConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import org.infinispan.commons.configuration.attributes.AttributeDefinition;
+import org.infinispan.commons.configuration.attributes.AttributeSet;
+import org.infinispan.commons.configuration.attributes.Matchable;
+
+/**
+ * Configures query options and defaults
+ */
+public class QueryConfiguration extends AbstractTypedPropertiesConfiguration implements Matchable<QueryConfiguration> {
+
+   public static final AttributeDefinition<Integer> DEFAULT_MAX_RESULTS = AttributeDefinition.builder(org.infinispan.configuration.parsing.Attribute.DEFAULT_MAX_RESULTS, 100).immutable().build();
```
Comment (tristantarrant) [inline core/src/main/java/org/infinispan/configuration/cache/QueryConfiguration.java:13]: For now, immutable is fine, but in the future we could turn this into a runtime-mutable attribute. Can you create a Jira for it ?

**Comment by tristantarrant** [comment]
We may want to add some query benchmarks to https://github.com/infinispan/infinispan-benchmarks

File: query/src/main/java/org/infinispan/query/impl/QueryDefinition.java (hunk 2 above)
Comment (fax4ever) [reply query/src/main/java/org/infinispan/query/impl/QueryDefinition.java]: good idea. changed it

**Comment by fax4ever** [comment]
thanks @tristantarrant for the review! 
I made the changes and created the follow up: https://issues.redhat.com/browse/ISPN-14232

**Comment by fax4ever** [comment]
@dvagnero I made my attempt to document the (restored) behavior and the new property here: [ISPN-14194 Document query default max results](https://github.com/infinispan/infinispan/pull/10378/commits/ada0b841ef8416ac6cbbd911e111163926713efc)

**Comment by fax4ever** [comment]
@wfink I don't know if want to take a look at it

File: documentation/src/main/asciidoc/topics/upgrading.adoc [hunk 4]
```diff
@@ -123,6 +123,12 @@ Any annotations will simply be copied as comments to the generated `.proto` file
 
 The `filesystem-realm`, an experimental feature which has never been documented, has been removed.
 
+== Query efficiency
+
+Restore the default query max results that used to be present in older Infinispan versions (prior of 12.1.5.Final)
+and provide a new configuration attribute to change this default, that is now restored to 100.
+Limiting the number of results returned by a query. Applies to indexed, non-indexed, and hybrid queries.
+It significantly improves performance of queries that don't have an explicit limit set.
```
Comment (dvagnero) [inline documentation/src/main/asciidoc/topics/upgrading.adoc:131]:
```suggestion
== Query efficiency
The `default-max-results` property that was present in Infinispan 12.1.5.Final and earlier versions has been restored.
This property limits the number of results returned by a query and significantly improves performance of queries that don't have an explicit limit set.
The `default-max-results` value is set to 100.
```

File: documentation/src/main/asciidoc/topics/ref_query_ickle.adoc [hunk 5]
```diff
@@ -40,6 +40,9 @@ conjunction with `Query.startOffset(long startOffset)` to achieve pagination of
 Query<Book> query = queryFactory.create("FROM org.infinispan.sample.Book WHERE title like '%clustering%' ORDER BY year").startOffset(20).maxResults(10)
 ----
 
+The default limit of `100` is applied if `maxResults` is not explicitly set to the query instance.
+You can change this default for all the queries by setting the cache property `query.default-max-result`.
```
Comment (dvagnero) [inline documentation/src/main/asciidoc/topics/ref_query_ickle.adoc:44]:
```suggestion
NOTE
====
If you don't explicitly set the `MaxResults` for a query instance, {brandname} limits the number of results returned by the query to `100`.
You can change the default limit by setting the `query.default-max-results` cache property.
====
```

File: documentation/src/main/asciidoc/topics/upgrading.adoc (hunk 4 above)
Comment (tristantarrant) [reply documentation/src/main/asciidoc/topics/upgrading.adoc:131]: Not quite @dvagnero : Infinispan never had a `default-max-results` property. It just had a default, unmodifiable value for it.

File: documentation/src/main/asciidoc/topics/ref_query_ickle.adoc (hunk 5 above)
Comment (fax4ever) [reply documentation/src/main/asciidoc/topics/ref_query_ickle.adoc]: Applied, thanks!

File: documentation/src/main/asciidoc/topics/upgrading.adoc (hunk 4 above)
Comment (fax4ever) [reply documentation/src/main/asciidoc/topics/upgrading.adoc]: Thanks @dvagnero, I merged them:

> The default query max results that was present in Infinispan 12.1.5.Final and earlier versions has been restored,
together with the fact that you can now also change this default by setting the `query.default-max-results` cache property.
Limiting the number of results returned by a query significantly improves performance of queries that don't have an explicit limit set.
The default value of `default-max-results` is 100.

@dvagnero what do you think of?

**Comment by fax4ever** [comment]
I applied the changed to the docs, thanks @dvagnero!

File: documentation/src/main/asciidoc/topics/upgrading.adoc (hunk 4 above)
Comment (dvagnero) [reply documentation/src/main/asciidoc/topics/upgrading.adoc]: Thanks! What about...
The default for maximum results returned by a query that was present in Infinispan 12.1.5.Final and earlier versions has been restored. You can now change the default limit by setting the `default-max-results` cache property. The default value of `default-max-results` is 100.
Limiting the number of results returned by a query significantly improves performance of queries that don't have an explicit limit set.

Comment (fax4ever) [reply documentation/src/main/asciidoc/topics/upgrading.adoc]: sounds great to me... I'm applying it. Thanks

File: core/src/main/resources/schema/infinispan-config-14.0.xsd (hunk 1 above)
Comment (wfink) [reply core/src/main/resources/schema/infinispan-config-14.0.xsd]: Looks good to me

**Comment by fax4ever** [comment]
Applied the last changes from @dvagnero.
I think that we can merge it if the CI passes.

**Comment by tristantarrant** [comment]
IN the previous run there were two thread leaks associated with indexing tests

**Comment by fax4ever** [comment]
> In the previous run there were two thread leaks associated with indexing tests:
> 
> ThreadLeakChecker – org.infinispan.query.api.ManualIndexingTest ThreadLeakChecker – org.infinispan.query.distributed.MassIndexingTest

@tristantarrant I don't know how the current changes may affect those tests. I think that we could have the same failures on the main branch.

**Comment by tristantarrant** [comment]
Merged, thanks

**Comment by fax4ever** [comment]
Thank you!

**Pull request merged**